### ISSUE TITLE
`ninja==1.11.1.2` build fails when built with `--no-binary=ninja` (or `--no-binary=:all:`)

### ISSUE THREAD

**Issue by TimotheusBachinger** [scikit-build/ninja-python-distributions#274]
Since `ninja==1.11.1.2`, we cannot built `numpy` anymore from source as it will fail with:
```
LookupError: https://files.pythonhosted.org/packages/b4/49/4f1a79f99f4c3eb5d22f943bba14832923bb44423254d5089d38a9f6da63/ninja-1.11.1.2.tar.gz (from https://pypi.org/simple/ninja/) (requires-python:>=3.7) is already being built: ninja>=1.8.2 from https://files.pythonhosted.org/packages/b4/49/4f1a79f99f4c3eb5d22f943bba14832923bb44423254d5089d38a9f6da63/ninja-1.11.1.2.tar.gz
```

`numpy` does not pin `ninja` in its build dependencies, which explains why the build is now failing:
https://github.com/numpy/numpy/blob/b1e6ccd14b173b922f5d6a11bb252f4141dc42cd/requirements/build_requirements.txt#L3 

Do you have any idea what causes this?

**Comment by johningve** [comment]
I'm also having issues with this new version. It seems like ninja is calling itself recursively:

![image](https://github.com/user-attachments/assets/cda90012-ead5-4b05-83a4-b56f3de5a950)

**Comment by mayeut** [comment]
@TimotheusBachinger, it's a bootstrapping issue https://github.com/scikit-build/cmake-python-distributions/issues/503#issuecomment-2191858783

ninja requires cmake and cmake is requiring ninja.
Installing the system cmake should solve the issue.

Any reason you're not getting / using binary wheels ? (specific platform maybe ?)

@johningve, this doesn't seem like the same issue, can you please open a different issue with the reproducer please ?

**Comment by TimotheusBachinger** [comment]
@mayeut thanks for pointing into that direction - this could really have caused the issue in our case (and due to heavy caching by bazel, to use it looked like the new ninja version...)

**Comment by TimotheusBachinger** [comment]
I am re-opening the issue again as the `cmake` theory was a dead end for us.
Our quick fix for now it to use a `constrain` when building from source: https://github.com/Checkmk/checkmk/commit/03164646ebcb5b7af7cf00a9c918770e4ed1cbf1

For us it is 100% the new version of `ninja==1.11.1.2` which leds to the build failure.

**Comment by mayeut** [comment]
@TimotheusBachinger,

In order to better understand the problem and provide a fix:
> Any reason you're not getting / using binary wheels ? Force build from source or specific platform maybe ?

**Comment by TimotheusBachinger** [comment]
@mayeut we're building for around 12 different platforms (debian, ubuntu, sles, centos, alma...) therefore we're building from source, yes (`--no-binary=":all:"`)

**Comment by mayeut** [comment]
ok, so I have an idea where the issue is coming from in that case. I will try something after day work to reproduce and hopefully fix this.
It's still a bootstrapping issue which this time should be handled by the new build system used by Ninja PyPI distribution.

**Comment by TimotheusBachinger** [comment]
@mayeut sounds good, thanks for the effort and no hurry. We're fine with the constraints workaround.

**Comment by harmstyler** [comment]
Is there any progress on this? Seems like there's an open Pull Request. Does this need extra help?

**Comment by mayeut** [comment]
This should be fixed with version 1.11.1.3 now available on PyPI
